### ISSUE TITLE
CORSMiddleware not work

### ISSUE THREAD

**Issue by LYH8112** [tiangolo/fastapi#1663]
Hi,
I searched the FastAPI documentation( https://fastapi.tiangolo.com/tutorial/cors/). But get errors:
Access to XMLHttpRequest at "http://127.0.0.1:8086/api" from origin "http://www.example.com" has been blocked by CORS policy:Response to preflight request doesn't pass access control check:No 'Access-Control-Allow-Origin' header is present on the requeste resource.

"http://127.0.0.1:8086/api" and "http://www.example.com" are in same pc.

from fastapi import FastAPI
from starlette.middleware.cors import CORSMiddleware

app = FastAPI()
app.add_middleware(
    CORSMiddleware,
    allow_origins=["*"],      
    allow_credentials=True,
    allow_methods=["*"], 
    allow_headers=["*"],
)

**Comment by phy25** [comment]
How do you make the request? Could you use the web devtools to see how the request look like (with HTTP headers)?

Also you may want to follow the issue template.

**Comment by LYH8112** [comment]
> How do you make the request? Could you use the web devtools to see how the request look like (with HTTP headers)?
> 
> Also you may want to follow the issue template.

var apiUrl = "http://127.0.0.1:8086/api"; 
var data-from = {  };
$.ajax({
        url: apiUrl,
        type: 'POST',
        data: JSON.stringify(data-from),
        contentType: 'application/json;charset=utf-8',
        dataType: 'json',                 
        cache: false,
        crossDomain: true,
        async: true,
       success: function (data) {
               alert(JSON.stringify(data));                       
       },
       error: function (data, err) {
             alert("fail " + JSON.stringify(data) + " " + JSON.stringify(err));
       }
});

Client:
Accept: */*
Accept-Encoding: gzip, deflate
Accept-Language: zh-CN,zh;q=0.9
User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/83.0.4103.97 Safari/537.36
Access-Control-Request-Headers: content-type
Access-Control-Request-Method: POST

Response:
HTTP/1.1 502 Fiddler

**Comment by SnkMaster** [comment]
You should add `CORSMiddleware` like this:

```python
app.add_middleware(
CORSMiddleware,
allow_origins=["*"], # Allows all origins
allow_credentials=True,
allow_methods=["*"], # Allows all methods
allow_headers=["*"], # Allows all headers
)
```

**Comment by darkclouder** [comment]
Please note that Safari does not support the wildcard `*` in some fields.
If wildcard is used in the middleware config, the middleware should explicitly send all values being used.
Source:
https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Allow-Headers#Browser_compatibility
https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Allow-Methods#Browser_compatibility

**Comment by BramVanroy** [comment]
I seem to be running into the same issue:

```python
app = FastAPI()
app.add_middleware(CORSMiddleware, allow_origins=["*"])

@app.post("/tokenize/", summary="Process batches of text", response_model=ResponseModel)
def tokenize(query: RequestModel):
    # do stuff
```
but after a POST request from my React front-end, console notifies me: 

```
Access to fetch at 'http://127.0.0.1:8000/tokenize/' from origin 'http://localhost:3000' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource. If an opaque response serves your needs, set the request's mode to 'no-cors' to fetch the resource with CORS disabled.
```

**Comment by ScrimForever** [comment]
@BramVanroy 
I used **_from starlette.middleware.cors import CORSMiddleware_** for workaround and works fine. If this can help you, let me know !

**Comment by ycd** [comment]
@ScrimForever was it an `import` issue in your case?

**Comment by ScrimForever** [comment]
> @ScrimForever was it an `import` issue in your case?

When i use fastapi CORSMiddleware, same as Bram, i receive error when use method POST, so, i change to starlette cors, and works. Just it.

**Comment by Mause** [comment]
That seems very strange, as fastapi doesn't have a cors implementation - it just exposes starlettes one

See https://github.com/tiangolo/fastapi/blob/master/fastapi/middleware/cors.py

**Comment by tricosmo** [comment]
I have the same issue here.  any updates, guys?

**Comment by ycd** [comment]
@tricosmo try importing from Starlette.

```python
from starlette.middleware.cors import CORSMiddleware
```

**Comment by tricosmo** [comment]
> @tricosmo try importing from Starlette.
> 
> ```python
> from starlette.middleware.cors import CORSMiddleware
> ```

Thanks @ycd , that is what I have. Plus as @Mause  mentioned about, it doesn't make sense. 
```Python
import logging

import uvicorn
from fastapi import APIRouter, Depends, FastAPI
from fastapi_aad_auth import AADAuth, AuthenticationState
from starlette.middleware.cors import CORSMiddleware

auth_provider = AADAuth()

logging.basicConfig(level="DEBUG")
router = APIRouter()


@router.get("/hello")
async def hello_world(
    auth_state: AuthenticationState = Depends(auth_provider.api_auth_scheme),
):
    print(auth_state)
    return {"hello": "world"}


app = FastAPI(
    title="fastapi_aad_auth test app",
    description="Testapp for Adding Azure Active Directory Authentication for FastAPI",
    openapi_url=f"/api/v1/openapi.json",
    docs_url="/api/docs",
    swagger_ui_init_oauth=auth_provider.api_auth_scheme.init_oauth,
    redoc_url="/api/redoc",
)

app.add_middleware(
    CORSMiddleware,
    allow_origins=["http://localhost:8001", "http://0.0.0.0:8001"],
    allow_credentials=True,
    allow_methods=["*"],
    allow_headers=["*"],
)


app.include_router(router)


if __name__ == "__main__":
    uvicorn.run(app, host="0.0.0.0", debug=True, port=8001, log_level="debug")

```
> Access to fetch at 'https://login.microsoftonline.com/c9da0d0e-f235-4e9a-b399-246a120ebd3e/oauth2/v2.0/token' from origin 'http://localhost:8001' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource. If an opaque response serves your needs, set the request's mode to 'no-cors' to fetch the resource with CORS disabled.

**Comment by Mause** [comment]
@tricosmo I think you have misunderstood the purpose of cors - you can only configure requests *to* your website, not *from*

**Comment by tricosmo** [comment]
@Mause Thanks, that is a bit of silly me.

**Comment by synchronizing** [comment]
As mentioned [here](https://github.com/tiangolo/fastapi/issues/133#issuecomment-673419315), one needs to use the `allow_all_origins` variable for `allow_origins=["*"]` to work. Effectively, this code works to expose _all_ origins correctly:

```python
app.add_middleware(CORSMiddleware, allow_all_origins=True)
```

**Comment by synchronizing** [comment]
My solution was to move my `app.add_middleware(CORSMiddleware, allow_origins=["*"])` to the bottom of all previous `app` configuration. I'm using `fastapi_versioning`, and it seems to break the CORS headers.

**Comment by ruoshui-git** [comment]
This is what I'm doing for now. It works for me:
```python
from fastapi import FastAPI
from starlette.middleware import Middleware
from starlette.middleware.cors import CORSMiddleware

origins = [
    "http://localhost:5000",
]

middleware = [
    Middleware(CORSMiddleware, allow_origins=origins)
]

app = FastAPI(middleware=middleware)
```

You can check [starlette](https://www.starlette.io/middleware/#corsmiddleware) for how to initialize middleware with Starlette's middlewares, and then put that in FastAPI.

**Comment by DocSavage** [comment]
Hit the same issue and the prior fixes aren't working.  The previous fixes didn't work for me.  I had to create an APIRoute and wire it into any APIRouter that needs CORS handling.

```
app = FastAPI()

# Handle CORS
class CORSHandler(APIRoute):
    def get_route_handler(self) -> Callable:
        original_route_handler = super().get_route_handler()

        async def preflight_handler(request: Request) -> Response:
            response = await original_route_handler(request)
            response.headers['Access-Control-Allow-Origin'] = '*'
            response.headers['Access-Control-Allow-Methods'] = 'POST, GET, DELETE, OPTIONS'
            response.headers['Access-Control-Allow-Headers'] = 'Authorization, Content-Type'
            return response

        return preflight_handler

router = APIRouter(route_class=CORSHandler)
...
app.include_router(router)
```

**Comment by DocSavage** [comment]
I noticed that OPTIONS requests were being denied until I explicitly added an endpoint handler like `@router.options() ...`

On further investigation, I saw that starlette allows a kind of path wildcard so any number of routes can be handled by one handler: https://stackoverflow.com/questions/63069190/how-to-capture-arbitrary-paths-at-one-route-in-fastapi

So rather than the above approach, which doesn't work without an explicit options handler, I simply do this now and it seems to work:
```
app = FastAPI()

# handle CORS preflight requests
@app.options('/{rest_of_path:path}')
async def preflight_handler(request: Request, rest_of_path: str) -> Response:
    response = Response()
    response.headers['Access-Control-Allow-Origin'] = ALLOWED_ORIGINS
    response.headers['Access-Control-Allow-Methods'] = 'POST, GET, DELETE, OPTIONS'
    response.headers['Access-Control-Allow-Headers'] = 'Authorization, Content-Type'
    return response

# set CORS headers
@app.middleware("http")
async def add_CORS_header(request: Request, call_next):
    response = await call_next(request)
    response.headers['Access-Control-Allow-Origin'] = ALLOWED_ORIGINS
    response.headers['Access-Control-Allow-Methods'] = 'POST, GET, DELETE, OPTIONS'
    response.headers['Access-Control-Allow-Headers'] = 'Authorization, Content-Type'
    return response
 ```
The odd thing is that the suggested CORSMiddleware seemed to work for some of my APIRouters but not others, and applying the middleware at the end of router initializations didn't seem to help.

**Comment by snow6oy** [comment]
Thanks @DocSavage this fix worked. But then after reading
[MDN](https://developer.mozilla.org/en-US/docs/Web/API/Request/credentials)
I realised that my JS client wasn't sending credentials properly. Fixing the client and re-testing I can see that either the DocSavage fix or CORSMiddleware will work as expected.

**Comment by edwardreed81** [comment]
I couldn't get CORS headers to work properly until I did what @DocSavage outlined. What's the story with the middleware?

**Comment by gocreating** [comment]
Same problem and all the proposed solutions above don't work.

**Comment by DocSavage** [comment]
A couple of people find that changing order of middleware can fix their problem, but I'm wondering if that only works if they aren't using APIRouter approaches (https://fastapi.tiangolo.com/tutorial/bigger-applications/#another-module-with-apirouter).  I believe I tried to reorder but couldn't get it to work when using many APIRouters with different modules. One would think that this would work but it didn't seem to work for endpoints in the APIRouter:

```python
app.include_router(foo.router, ...)
app.add_middleware(CORSMiddleware, ...)
```
Can someone verify the above works for them?

**Comment by gocreating** [comment]
@DocSavage I think it makes more sense to put middlewares before routers. At least that works for me.

**Comment by gocreating** [comment]
OMG
The working code yesterday gets CORS error back today...

**Comment by merrcury** [comment]
Hi, I deployed an app using FastAPI, with CORS set to my frontend IP only but I am still able to call it from another IPs as well.

**Comment by antwigambrah** [comment]
Hi , if anyone is still having the CORSMiddleware issue,this is what works for me 

```
from starlette.middleware.cors import CORSMiddleware
from starlette.middleware import Middleware

app = FastAPI()

your routes 

origins=[*]
app = CORSMiddleware(
    app=app,
    allow_origins=orgins,
    allow_credentials=True,
    allow_methods=["*"],
    allow_headers=["*"],
)
```

**Comment by raghuveersunkara** [comment]
I am still having issues with API routers. It is working with the methods in main.py but router endpoints still fail with CORS issue. Is it working with anyone else? On `http://localhost:8000` endpoint, all routes work fine, but this fails when deployed to AWS Lambda -> API Gateway. Can someone help? My `main.py` file is below
```
from fastapi import FastAPI
from mangum import Mangum
from starlette.middleware.cors import CORSMiddleware

from .api.app_v1.api import router as api_router
import json


origins = ["*"]


app = FastAPI() #(middleware=middleware)

app.add_middleware(
    CORSMiddleware,
    allow_origins=origins,
    allow_credentials=True,
    allow_methods=["*"],
    allow_headers=["*"],
)

app.include_router(api_router, prefix="/api/v1")

handler = Mangum(app)


@app.get("/")
async def root():
    """
    Base root method
    """
    return {
        'statusCode': 200,
        'headers': {
            'Access-Control-Allow-Headers': 'Content-Type,X-Amz-Date,Authorization,X-Api-Key,X-Amz-Security-Token',
            'Access-Control-Allow-Origin': '*',
            'Access-Control-Allow-Methods': 'DELETE, GET, HEAD, OPTIONS, PATCH, POST, PUT'
        },
        'body': json.dumps('Hello World!')
    }
```

**Comment by manoadamro** [comment]
this title is 100% accurate. I'm using a single middleware:

```
app = FastAPI(middlewares=[
    Middleware(
        CORSMiddleware,
        allow_origins=["*"],
        allow_credentials=False,
        allow_methods=["*"],
        allow_headers=["*"]
    )
])
```

and it's about as lenient as it could possibly be, yet I can't make a single request to the API. I'm no expert in this kind of stuff but perhaps someone could explain why the default behaviour is not being able to hit your API at all?

**Comment by dazza-codes** [comment]
With CORS middleware enabled, any prefetch OPTIONS requests need to include an extra header to match the method of the route, e.g.

```
$ curl --location --request OPTIONS 'http://127.0.0.1:8080/a_post_route' --header 'Origin: http://example.com'
{"detail":"Method Not Allowed"}
$ curl --location --request OPTIONS 'http://127.0.0.1:8080/a_post_route' --header 'Origin: http://example.com' --header 'Access-Control-Request-Method: POST' 
OK
```

The `/a_post_route` is decorated with `app.post(...)` and it does not support any methods other than POST.

**Comment by rayvikram** [comment]
Just a note. Order of middleware matters, If you are having multiple middlewares, the CorsMiddleware should be the last one.

**Comment by callmeshades** [comment]
If anyone else has this issue make sure to follow as stated in the [documentation](https://fastapi.tiangolo.com/tutorial/cors/?h=cors).

> allow_credentials - Indicate that cookies should be supported for cross-origin requests. Defaults to False. 
> Also, allow_origins cannot be set to ['*'] for credentials to be allowed, origins must be specified.

In this case to allow the CORS middleware.

```
from starlette.middleware.cors import CORSMiddleware

app.add_middleware(
    CORSMiddleware,
    allow_origins=["http://localhost", "http://localhost:3000", "http://localhost:8000"],
    allow_credentials=True,
    allow_methods=["*"],
    allow_headers=["*"],
)
```

Once the origins are provided it works! If you're mixing 127.0.0.1 and localhost, don't. Unless you add them both to the origins, make sure to stick with one or the other.

**Comment by dorinclisu** [comment]
I have another problem with CORSMiddleware, but since the cause of this issue is still not known, I'm gonna write here because there already seems to be large number of open issues.

Basically, it all works (with swagger and with JS app, locally and on server) except when an exception is raised in the path operation function. By default, fastapi returns a 500 response. All good up to this point.

Here's when things get strange: while the 500 response shows up on swagger as Internal Server Error (which is good), in the JS app the response never comes, instead it's flagged as a failed CORS check.

Since the swagger app code comes from the same origin, it's understandable the browser doesn't check the CORS headers. The question is why the CORS middleware fails to validate the origin, causing the 500 response to never reach the browser.

**Comment by gitblit** [comment]
Interesting question @dorinclisu.  I don't know the answer, but I can share that I've seen this exact browser behavior with Lambdas (regardless of tech choice) exposed via API Gateway - browser indicates CORS error on endpoint 500s.  Perhaps there is a common CORS setup issue that triggers this browser behavior.  Not helpful, I know. 🤷‍♂️

**Comment by matiasbeckerle** [comment]
Thank you all for the comments -- it helped us to continue making progress and the [_set CORS headers response_](https://github.com/tiangolo/fastapi/issues/1663#issuecomment-763188503) worked for us as well.

But then we decided to give the middleware a new try and debug it to troubleshoot what was going on in our environments. In the source code there is a couple of references to the `Origin` [header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Origin) which triggered me to review if we were sending it or not. While running the api locally I saw in the browser request headers there was not an `Origin` header.

So we ended up removing all the things we tried and we started from the top. Our api ended up having only:

```python
from fastapi import FastAPI
from fastapi.middleware.cors import CORSMiddleware

app = FastAPI()

app.add_middleware(
    CORSMiddleware, allow_origins=["*"], allow_credentials=True, allow_methods=["*"], allow_headers=["*"]
)
```

And as long as clients sends the `Origin` header the api response will contain the headers we were missing before:

```
access-control-allow-credentials true
access-control-allow-origin *
content-length 48
content-type application/json
date Thu, 20 May 2021 14:45:43 GMT
server uvicorn
```

This also works for the `OPTIONS` requests as well.

Extra: the associated line from the source code is cors.py:75

```python
origin = headers.get("origin")

if origin is None:
  await self.app(scope, receive, send)
  return
```

**TL;DR: Check if you are including the `Origin` header in your requests.**

**Comment by fquirin** [comment]
Just something I wanted to mention. I was looking for the error for about an hour, checking all the proposals here, checking the 'Origin' header and all settings seemed to be ok. Then finally I found the problem. In my function I was using `settings.origins` which is created from a config file and at some point changed from `'*'` to `['*']`:
```
app.add_middleware(
    CORSMiddleware, allow_origins=[settings.origins], allow_credentials=True, allow_methods=["*"], allow_headers=["*"]
)
```

So ultimately I ended up with `allow_origins=[['*']]` instead of `allow_origins=['*']`, no error was ever thrown though :unamused: 
Seems like it was simply ignored :man_shrugging: .
Maybe there could be some form of validity check?

**Comment by slavaGanzin** [comment]
It would be awesome to write in documentation: THIS FEATURE IS NOT WORKING. So, anyone wouldn't assume it's working. As it should.

**Comment by fquirin** [comment]
well, it does work, I'm using it exactly as described, ... it just doesn't tell you when your settings are invalid :man_shrugging:

**Comment by kevinadhiguna** [comment]
> Just something I wanted to mention. I was looking for the error for about an hour, checking all the proposals here, checking the 'Origin' header and all settings seemed to be ok. Then finally I found the problem. In my function I was using `settings.origins` which is created from a config file and at some point changed from `'*'` to `['*']`:
> 
> ```
> app.add_middleware(
>     CORSMiddleware, allow_origins=[settings.origins], allow_credentials=True, allow_methods=["*"], allow_headers=["*"]
> )
> ```
> 
> So ultimately I ended up with `allow_origins=[['*']]` instead of `allow_origins=['*']`, no error was ever thrown though unamused
> Seems like it was simply ignored man_shrugging .
> Maybe there could be some form of validity check?

Mine did not work : 

```python
from fastapi import FastAPI
from fastapi.middleware.cors import CORSMiddleware

app = FastAPI()

# CORS configuration
origins = os.environ['CORS_ALLOW_ORIGINS'].split(',') if os.environ['CORS_ALLOW_ORIGINS'] else None

app.add_middleware(
    CORSMiddleware,
    allow_origins=origins,
    allow_credentials=True,
    allow_methods=["*"],
    allow_headers=["*"],
)
```

Let's say my environment variable (.env) is : 

```
CORS_ALLOW_ORIGINS=http://192.168.1.4:3000
```

**Comment by fquirin** [comment]
Did you try hard-coding `allow_origins=["http://192.168.1.4:3000"]` or `["*"]` just to make sure?

**Comment by maulberto3** [comment]
Hi, I am following the simplest CORS FastAPI, it worked, but now, without ANY change to code, a few weeks, now it doesn't...

**Comment by taghizad3h** [comment]
In my case it was due to nginx misconfiguration that forwarded my request to a wrong port.
Any way currently CORSMiddleware is working fine.

**Comment by nefarioustim** [comment]
I am seeing similar behaviour to @maulberto3 in that, when my POST data is incorrect, the CORS headers are dropped from the response. Here's a curl call that shows my output:

```shell
$ curl -iX 'POST' \
  'https://myapiendpointhere.com' \
  -H 'accept: application/json' \
  -H 'Authorization: Bearer mysupersecretjwthere' \
  -H 'Content-Type: application/json' \
  -d '{
  "foo": "bar"
}'

HTTP/1.1 422 Unprocessable Entity
date: Wed, 25 Aug 2021 08:18:52 GMT
server: uvicorn
content-length: 89
content-type: application/json

{"detail":[{"loc":["body","title"],"msg":"field required","type":"value_error.missing"}]}%    
```

**Comment by Aradhya-Tripathi** [comment]
Reading through the solutions as it turns out it's just about the order of the middleware, If you have set up custom middleware just make sure to add the cors middleware at the end.

Ps. fast API just exposes starlette cors

**Comment by DocSavage** [comment]
@Aradhya-Tripathi There have been a number of people commenting it's just about the order of the middleware as if all the people having trouble never grokked that idea.  If you look back at my original fix, I made it clear that I tried that possible fix and it didn't work under all scenarios. Others also mention they've tried that solution as well. Just because someone can get it to work for their configuration doesn't invalidate the bug reports from others.

**Comment by Aradhya-Tripathi** [comment]
> mention



> @Aradhya-Tripathi There have been a number of people commenting it's just about the order of the middleware as if all the people having trouble never grokked that idea. If you look back at my original fix, I made it clear that I tried that possible fix and it didn't work under all scenarios. Others also mention they've tried that solution as well. Just because someone can get it to work for their configuration doesn't invalidate the bug reports from others.

I do understand that but this worked for my configuration, this does seem to be a valid bug not disregarding anything 😟

**Comment by dorinclisu** [comment]
Just FYI, after plenty of observations and debugging I noticed that the CORS documentation at https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS is no longer true / complete.

Browsers are now more strict than the specification. For example, a few pain points for Chrome:
1. It can make OPTIONS (pre-flight) requests before each GET request (could not find the exact conditions in which it doesn't).
2. It requires the CORS headers be set also on GET responses, not just the OPTIONS responses (this can have significant implications if you are relying on an external middleware i.e. AWS API Gateway or nginx to handle the CORS).
3. Developer tools no longer show the OPTIONS requests, instead they simply show the GET request failed with a message in the console log that does not always reflect the reality. This makes debugging the exact failure point very tricky.

**Comment by dorinclisu** [comment]
> > Just FYI, after plenty of observations and debugging I noticed that the CORS documentation at https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS is no longer true / complete.
> > Browsers are now more strict than the specification. For example, a few pain points for Chrome:
> > 
> > 1. It can make OPTIONS (pre-flight) requests before each GET request (could not find the exact conditions in which it doesn't).
> > 2. It requires the CORS headers be set also on GET responses, not just the OPTIONS responses (this can have significant implications if you are relying on an external middleware i.e. AWS API Gateway or nginx to handle the CORS).
> > 3. Developer tools no longer show the OPTIONS requests, instead they simply show the GET request failed with a message in the console log that does not always reflect the reality. This makes debugging the exact failure point very tricky.
> 
> So do we need some update to this feature.

I don't know for sure. In a basic scenario, FastAPI seems to work as is (it already returns CORS headers on most responses) but the point is that depending on the exact deployment method, the tricky bits need to be kept in mind.

BTW, I just realized that I probably found the cause of my own problem posted here a few months ago:
> I have another problem with CORSMiddleware, but since the cause of this issue is still not known, I'm gonna write here because there already seems to be large number of open issues.
> 
> Basically, it all works (with swagger and with JS app, locally and on server) except when an exception is raised in the path operation function. By default, fastapi returns a 500 response. All good up to this point.
> 
> Here's when things get strange: while the 500 response shows up on swagger as Internal Server Error (which is good), in the JS app the response never comes, instead it's flagged as a failed CORS check.
> 
> Since the swagger app code comes from the same origin, it's understandable the browser doesn't check the CORS headers. The question is why the CORS middleware fails to validate the origin, causing the 500 response to never reach the browser.

So, I think this is caused by pain point no. 2, more exactly because FastAPI middleware does not send CORS headers on a 500 response, at least with the default exception handler. The solution is obvious, to modify the middleware always send the CORS headrs..

**Comment by yhatpub** [comment]
If anyone wants an example, I just open-sourced a server I was working on and CORS works as can be seen at https://yhat.pub

here is where the cors middleware is
https://github.com/yhatpub/yhat_webserver/blob/main/app/api.py#L63

**Comment by fquirin** [comment]
> well, it does work, I'm using it exactly as described, ... it just doesn't tell you when your settings are invalid man_shrugging

After I fixed my syntax error I never had any issues again. Here is the code example:
https://github.com/SEPIA-Framework/sepia-stt-server/blob/master/src/server.py#L16

**Comment by reinaldo-gomes** [comment]
> @tricosmo I think you have misunderstood the purpose of cors - you can only configure requests _to_ your website, not _from_

I spent several hours trying to understand why I was getting CORS errors, even though I was setting allowed origins for CORSMiddleware on my FastAPI backend, only to realize that I had to set it on my authorization server (keycloak) instead... Thank you.

**Comment by realmhamdy** [comment]
I wrestled with this for a couple of hours and honestly I'll just switch back to Django/DRF fuck these kids. It's a new project with a single endpoint so that's something I can afford.

**Comment by dorinclisu** [comment]
> I kindly ask you to let me know what I'm doing wrong, I checked the documentation many times and still can't figure it out.
> Thank you all in advance.

Are the requests serviced or not? We cannot see if the status code is 200, but if yes then there's no problem. Did you read https://github.com/tiangolo/fastapi/issues/1663#issuecomment-982582094 ? This is normal browser behavior.

**Comment by PaverStrike** [comment]
@dorinclisu thank you for the reply.

Yes, all the requests are 200. I can fetch data and nothing breaks.
Just wondering if seeing all these requests being made by the `webapp` subdomain to the `api` subdomain are actually expected.

Thank you again

**Comment by nicholasbleo** [comment]
I was able to get this to work by adding `"http://127.0.0.1:8080"` (instead of only `"http://localhost:8080"`) to the list of allowed origins, using starlette's `CORSMiddleware` module instead of FastAPI's, and using the middleware options as a parameter instead of using `app.add_middleware()`.

I was also using python's http.server to serve the front-end, and I had to make sure to bind that to 127.0.0.1 like so:
`python -m http.server 8080 --bind 127.0.0.1`

Like so:

```
from starlette.middleware.cors import CORSMiddleware
from fastapi.middleware import Middleware
```

```
origins = [
    "http://localhost",
    "http://localhost:8000"
    "http://localhost:8080",
    "http://127.0.0.1:8080"
]

middleware = [ Middleware(
    CORSMiddleware,
    allow_origins=origins,
    allow_credentials=True,
    allow_methods=["*"],
    allow_headers=["*"],
) ]

app = FastAPI(middleware=middleware)
```

I had to visit `127.0.0.1:8080` specifically - not `localhost:8080`

**Comment by Butterweck** [comment]
had the same issue. cors didnt seem to work for one specific post request.

turn out, the user was sending 0 for one body parameter. wich lead to a division by zero exception and a 500 error.
with the 500 error, the headers seem to have been removed, which in turn led to a cors error message and the user to believe, it was a cors issue, when in fact it was the 0 case not being handled correctly.

**Comment by Pearseak** [comment]
> 

I lost 3 days to this...THIS

**Comment by Pearseak** [comment]
> One little thing that could be causing some heartburn... **trailing slash**. Erase it.
> 
> good -> `allow_origins=[https://www.example.com]` bad -> `allow_origins=[https://www.example.com/]`

I lost 3 days to this...THIS

**Comment by Butterweck** [comment]
> I lost 3 days to this...THIS
here too ^^

**Comment by michjnich** [comment]
Well, thanks to @dorinclisu for solving my problem. A backend issue had crept in but the 500 error it created erroneously presented as a CORS error because the headers were either dropped or never added.I was so focused on trying to work out what was wrong with the CORS setup I'd not spotted the other error, and once that was fixed all was good again!

**Comment by Aravind-psiog** [comment]
> I am still having issues with API routers. It is working with the methods in main.py but router endpoints still fail with CORS issue. Is it working with anyone else? On `http://localhost:8000` endpoint, all routes work fine, but this fails when deployed to AWS Lambda -> API Gateway. Can someone help? My `main.py` file is below
> 
> ```
> from fastapi import FastAPI
> from mangum import Mangum
> from starlette.middleware.cors import CORSMiddleware
> 
> from .api.app_v1.api import router as api_router
> import json
> 
> 
> origins = ["*"]
> 
> 
> app = FastAPI() #(middleware=middleware)
> 
> app.add_middleware(
>     CORSMiddleware,
>     allow_origins=origins,
>     allow_credentials=True,
>     allow_methods=["*"],
>     allow_headers=["*"],
> )
> 
> app.include_router(api_router, prefix="/api/v1")
> 
> handler = Mangum(app)
> 
> 
> @app.get("/")
> async def root():
>     """
>     Base root method
>     """
>     return {
>         'statusCode': 200,
>         'headers': {
>             'Access-Control-Allow-Headers': 'Content-Type,X-Amz-Date,Authorization,X-Api-Key,X-Amz-Security-Token',
>             'Access-Control-Allow-Origin': '*',
>             'Access-Control-Allow-Methods': 'DELETE, GET, HEAD, OPTIONS, PATCH, POST, PUT'
>         },
>         'body': json.dumps('Hello World!')
>     }
> ```

Everything looks good except for the header. Headers in FastAPI are different compared to lambda,s. This is what I did and it is working

```@app.get('/')
def root():
    content = {"message": "Hello From AWS!"}
    headers = {'Access-Control-Allow-Headers': 'Content-Type',
               'Access-Control-Allow-Origin': '*',
               'Access-Control-Allow-Methods': '*'}
    return JSONResponse(content=content, headers=headers)```

**Comment by hypercubex** [comment]
ran into the same issue, after debugging for a while, the CORS middleware will only proceed requests when both the `Origin` and `Access-Control-Request-Method` header exist

**Comment by nickgieschen** [comment]
FWIW, I started getting this once I upgraded to ngrok 3. Once I downgraded to ngrok 2.3.40, everything was working again.

**Comment by kiru** [comment]
I tried everything above, none of them worked. This solution worked for me: 
https://stackoverflow.com/a/65994876/5553845

**Comment by cointreau** [comment]
@kiru I finally solved thanks to you!! I really appreciate it

**Comment by k2ev** [comment]
> nd why I was getting CORS errors, even though I was setting allowed origins for CORSMiddleware on my FastAPI backend, only to realize that I had to set it on my authorization server (keycloak) instead... Thank you.

how did you solve it?

**Comment by cointreau** [comment]
@k2ev 
I just followed Sam_Ste's answer in the stackoverflow link kiru refered.

**Comment by p0lygun** [comment]
after wasting a lot of time i found the error with mine, it was a comma `,` 
my origins were defined as 
```py
origins = [ ... ],
```
which made it a tuple and `CORSMiddleware` takes a `List[str]`
[see](https://github.com/encode/starlette/blob/e3de71f1433568a3f7ac3dc69ef86cf4f72f5651/starlette/middleware/cors.py#L17)

IMO starlette should raise a error if that happens

**Comment by fquirin** [comment]
> IMO starlette should raise a error if that happens

I support this. Similar issue about a year ago: https://github.com/tiangolo/fastapi/issues/1663#issuecomment-873736512

**Comment by sl00k** [comment]
any solution to this? None of the work arounds are working for me :D

**Comment by JarroVGIT** [comment]
Can you be more specific? There are numerous issues with different causes discussed in this thread. Best would be to create a minimal reproducible example that demonstrates your configuration and error that you run into and open a new issue.

**Comment by archienapampe** [comment]
> Hit the same issue and the prior fixes aren't working. I had to create an APIRoute and wire it into any APIRouter that needs CORS handling. Here's the documentation pertinent to this approach: https://fastapi.tiangolo.com/advanced/custom-request-and-route/ But for some reason I'm getting all OPTIONS disallowed with a 405 method not allowed, so trying to see how to prevent that.
> 
> ```
> app = FastAPI()
> 
> # Handle CORS
> class CORSHandler(APIRoute):
>     def get_route_handler(self) -> Callable:
>         original_route_handler = super().get_route_handler()
> 
>         async def preflight_handler(request: Request) -> Response:
>             if request.method == 'OPTIONS':
>                 response = Response()
>                 response.headers['Access-Control-Allow-Origin'] = '*'
>                 response.headers['Access-Control-Allow-Methods'] = 'POST, GET, DELETE, OPTIONS'
>                 response.headers['Access-Control-Allow-Headers'] = 'Authorization, Content-Type'
>             else:
>                 response = await original_route_handler(request)
> 
>         return preflight_handler
> 
> router = APIRouter(route_class=CORSHandler)
> ...
> app.include_router(router)
> ```

Many thanks, I added some more http method I need for handling, it works good and resolved my problem with cors error when I implemented blueprints in the project!

**Comment by StankovskiA** [comment]
After spending many hours, I realized the reason why i was getting the CORS error was that my web service was failing ( I was casting to int in cases where the data was empty). 
Please test your app locally!

**Comment by artemonsh** [comment]
Please check that you are on the "localhost:8080" in your browser if you use origins = ["http://localhost:8080"] 
and  on "127.0.0.1:8080"  if you use origins = ["http://127.0.0.1:8080"]

**Comment by solarjoe** [comment]
To all the people that solved their problem by moving the configuration of the `CORSMiddleware` to the bottom of the code - after the endpoints - you might have made the same mistake as me.

Using lots of copypasted code, I was spinning up another FastAPI instance somewhere where CORS was not configured, like this:

    app = FastAPI()
    
    app.add_middleware(
        CORSMiddleware,
        allow_origins=['*']
    )

    ... much later somewhere within lots of green code ...
    app = FastAPI()

So, when I moved the the CORS configuration to the end of the code it was working :)

**Comment by jasstionzyf** [comment]
> 

Totally agree with you, if your service throw exceptions and not catch, then it will interrupt fastAPI flow, then the CORSMiddleware
will not executed.

**Comment by bhaveshachhada** [comment]
> > I am still having issues with API routers. It is working with the methods in main.py but router endpoints still fail with CORS issue. Is it working with anyone else? On `http://localhost:8000` endpoint, all routes work fine, but this fails when deployed to AWS Lambda -> API Gateway. Can someone help? My `main.py` file is below
> > ```
> > from fastapi import FastAPI
> > from mangum import Mangum
> > from starlette.middleware.cors import CORSMiddleware
> > 
> > from .api.app_v1.api import router as api_router
> > import json
> > 
> > 
> > origins = ["*"]
> > 
> > 
> > app = FastAPI() #(middleware=middleware)
> > 
> > app.add_middleware(
> >     CORSMiddleware,
> >     allow_origins=origins,
> >     allow_credentials=True,
> >     allow_methods=["*"],
> >     allow_headers=["*"],
> > )
> > 
> > app.include_router(api_router, prefix="/api/v1")
> > 
> > handler = Mangum(app)
> > 
> > 
> > @app.get("/")
> > async def root():
> >     """
> >     Base root method
> >     """
> >     return {
> >         'statusCode': 200,
> >         'headers': {
> >             'Access-Control-Allow-Headers': 'Content-Type,X-Amz-Date,Authorization,X-Api-Key,X-Amz-Security-Token',
> >             'Access-Control-Allow-Origin': '*',
> >             'Access-Control-Allow-Methods': 'DELETE, GET, HEAD, OPTIONS, PATCH, POST, PUT'
> >         },
> >         'body': json.dumps('Hello World!')
> >     }
> > ```
> 
> Everything looks good except for the header. Headers in FastAPI are different compared to lambda,s. This is what I did and it is working
> 
> ```
> def root():
>     content = {"message": "Hello From AWS!"}
>     headers = {'Access-Control-Allow-Headers': 'Content-Type',
>                'Access-Control-Allow-Origin': '*',
>                'Access-Control-Allow-Methods': '*'}
>     return JSONResponse(content=content, headers=headers)```
> ```

What did you exactly change here ?

**Comment by fquirin** [comment]
... and why would you overwrite the headers again? They are already defined via the `add_middleware` function 🤷‍♂️. This only makes sense if you want to explicitly limit the allowed headers to 'Content-Type' **only**.

**Comment by Aravind-psiog** [comment]
@bhaveshachhada refer to my repo https://github.com/Aravind-psiog/Monitor_API
And hope you have enabled cors on AWS API gateway and deployed it. If not, enable and re-deploy the API.

**Comment by bhaveshachhada** [comment]
> @bhaveshachhada refer to my repo https://github.com/Aravind-psiog/Monitor_API And hope you have enabled cors on AWS API gateway and deployed it. If not, enable and re-deploy the API.

@Aravind-psiog My API Gateway uses proxy integration, which means it doesn't change the response given by The lambda (FastAPI server) in any way. My app was working fine with Flask, which means the app should work with fastapi, without doing any changes on the API gateway level. Moreover, for me, creating a custom middleware worked.
```
@app.middleware("http")
    async def cors_handler(request: Request, call_next):
        response: Response = await call_next(request)
        response.headers['Access-Control-Allow-Credentials'] = 'true'
        response.headers['Access-Control-Allow-Origin'] = os.environ.get('allowedOrigins')
        response.headers['Access-Control-Allow-Methods'] = '*'
        response.headers['Access-Control-Allow-Headers'] = '*'
        return response
```
But the question is, this is just a workaround. Why adding the CORSMiddleware didn't work ?

**Comment by Joshuamjv2** [comment]
If you are deploying your API on AWS Lambda and using the URL provided by AWS Lambda, set your cors configuration in the Lambda console. Don't configure CORS with FastAPI. That's what worked for me.

**Comment by NatanSiilva** [comment]
I managed to solve it after a lot of struggle hehe. Cors with fastapi, the url of the client that can access the end-point of the api must be declared and also in the response of each request it must be declared which client can receive the response. Unlike django, which has a setup that abstracts all this logic, making it necessary to declare the client's url only once. So I created a middleware to add the header in the requests responses.

from fastapi import FastAPI
from fastapi.middleware.cors import CORSMiddleware

from .routes import api

app = FastAPI(
    openapi_url="/openapi.json",
    docs_url="/docs",
    redoc_url="/redoc",
)


@app.middleware("http")
async def add_cors_headers(request, call_next):
    response = await call_next(request)
    response.headers["Access-Control-Allow-Origin"] = "https://www.teste.com.br, http://localhost:3000"
    response.headers["Access-Control-Allow-Credentials"] = "true"
    response.headers["Access-Control-Allow-Methods"] = "GET, POST, PUT, DELETE, OPTIONS"
    response.headers["Access-Control-Allow-Headers"] = "Content-Type, Authorization"
    return response

origins = [
    "https://teste.com.br",
    "http://localhost:3000",
]

app.add_middleware(
    CORSMiddleware,
    allow_origins=origins,
    allow_credentials=True,
    allow_methods=["*"],
    allow_headers=["*"],
)

app.include_router(api)
`
`
